flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java

```diff
@@ -0,0 +1,297 @@
+					}
+				}
+			} else {
+				// Received an old confirmation call
+				if (!leaderSessionID.equals(this.issuedLeaderSessionID)) {
+					if (LOG.isDebugEnabled()) {
+						LOG.debug("Receive an old confirmation call of leader session ID {}, " +
+							"current issued session ID is {}", leaderSessionID, issuedLeaderSessionID);
+					}
+				} else {
+					LOG.warn("The leader session ID {} was confirmed even though the " +
+						"corresponding JobManager was not elected as the leader.", leaderSessionID);
+				}
+			}
+		}
+	}
+
+	@Override
+	public boolean hasLeadership(@Nonnull UUID leaderSessionId) {
+		synchronized (lock) {
+			if (running) {
+				return leaderElectionDriver.hasLeadership() && leaderSessionId.equals(issuedLeaderSessionID);
+			} else {
+				if (LOG.isDebugEnabled()) {
+					LOG.debug("hasLeadership is called after the service is stopped, returning false.");
+				}
+				return false;
+			}
+		}
+	}
+
+	/**
+	 * Returns the current leader session ID or null, if the contender is not the leader.
+	 *
+	 * @return The last leader session ID or null, if the contender is not the leader
+	 */
+	@VisibleForTesting
+	@Nullable
+	public UUID getLeaderSessionID() {
+		return confirmedLeaderSessionID;
+	}
+
+	@GuardedBy("lock")
+	private void confirmLeaderInformation(UUID leaderSessionID, String leaderAddress) {
+		confirmedLeaderSessionID = leaderSessionID;
+		confirmedLeaderAddress = leaderAddress;
+		leaderElectionDriver.writeLeaderInformation(
+			LeaderInformation.known(confirmedLeaderSessionID, confirmedLeaderAddress));
+	}
+
+	@GuardedBy("lock")
+	private void clearConfirmedLeaderInformation() {
+		confirmedLeaderSessionID = null;
+		confirmedLeaderAddress = null;
+	}
+
+	@Override
+	@GuardedBy("lock")
+	public void onGrantLeadership() {
+		synchronized (lock) {
+			if (running) {
+				issuedLeaderSessionID = UUID.randomUUID();
+				clearConfirmedLeaderInformation();
+
+				if (LOG.isDebugEnabled()) {
+					LOG.debug(
+						"Grant leadership to contender {} with session ID {}.",
+						leaderContender.getDescription(),
+						issuedLeaderSessionID);
+				}
+
+				leaderContender.grantLeadership(issuedLeaderSessionID);
+			} else {
+				if (LOG.isDebugEnabled()) {
+					LOG.debug("Ignoring the grant leadership notification since the {} has " +
+						"already been closed.", leaderElectionDriver);
+				}
+			}
+		}
+	}
+
+	@Override
+	@GuardedBy("lock")
+	public void onRevokeLeadership() {
+		synchronized (lock) {
+			if (running) {
+				if (LOG.isDebugEnabled()) {
+					LOG.debug(
+						"Revoke leadership of {} ({}@{}).",
+						leaderContender.getDescription(),
+						confirmedLeaderSessionID,
+						confirmedLeaderAddress);
+				}
+
+				issuedLeaderSessionID = null;
+				clearConfirmedLeaderInformation();
+
+				leaderContender.revokeLeadership();
+
+				// Clear the old leader information on the external storage
+				leaderElectionDriver.writeLeaderInformation(LeaderInformation.empty());
```

Comment:
The reason why I want to do this is to process the following situation.
* Old leader lost leadership and new leader is not launched very fast(maybe 120s or longer).
* Old leader should clear the leader information on ConfigMap so that the listener could know this event.

For Kubernetes, it could work. Because no new leader has been elected and the lock identity annotation is still the old one.

Reply:
ok, this makes sense. Thanks for the explanation.